File: 0090-subsets-ii/0090-subsets-ii.cpp
class Solution {
public:
    void explore(vector<int>&nums, int i, vector<vector<int>>&ans, vector<int>&curr){
        if(i==nums.size()){
            ans.push_back(curr);
            return;
        }
        int val = nums[i];
        curr.push_back(nums[i]);
        explore(nums, i+1, ans, curr);
        curr.pop_back();
        while(i<nums.size() and nums[i]==val) i++;
        explore(nums, i, ans, curr);
    }
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>>ans;
        vector<int>curr;

        explore(nums, 0, ans, curr);
        return ans;
    }
};

File: 0090-subsets-ii/0090-subsets-ii.py
class Solution:
    
    def helper(self, nums, i, li, ans):
        if i==len(nums):
            return
        li.append(nums[i])
        self.helper(nums, i+1, li, ans)
        ans.append(li[:])
        li.pop()

        while i<len(nums)-1 and nums[i]==nums[i+1]:
            i+=1
        self.helper(nums, i+1, li, ans)

    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        li = list()
        ans = list()
        ans.append([])
        nums.sort()
        self.helper(nums, 0, li, ans)
        # print(ans)
        return ans
        
        